### ISSUE TITLE
Feature Request: CSS Lint support

### ISSUE THREAD

**Issue by nzakas** [bdkjones/CodeKit-1-Old#110]
The JavaScript linting is a great help in CodeKit, it would be great to add CSS Lint support to check CSS syntax as well.

The project is open source and located here:
http://github.com/stubbornella/csslint

You can see it in action here:
http://www.csslint.net

**Comment by feelepxyz** [comment]
I second that motion!

PS, this app is shaping up to be the bomb!

**Comment by bdkjones** [comment]
Yea... CSSLint isn't really appropriate for CodeKit. 

If you're doing things right, your CSS is generated by either the LESS, SASS, SCSS or STYLUS compiler. It's guaranteed to be valid. Plus, even if you ran CSSLint on the outputted CSS, you'd have no way to "fix" the issue. Modifying the CSS file would simply result in an overwrite the next time you compiled the source Less file. 

Next, most people will choose to have the compilers minify the CSS they spit out. Does CSSLint work on minified files, or does it throw a bunch of "; expected" or "whitespace needed" errors? (I've never used it.)

On top of that, CodeKit does not watch or care about CSS files. This makes life simple. If we were to add CSSLint support, then the app would need to know about CSS files.

I have not ruled this idea out in a limited fashion --- perhaps running the outputted CSS through lint as you compile a Less/Sass/Stylus file. But let's just say this one is low on my priority list right now.

**Comment by nzakas** [comment]
Some things you should know about CSS Lint:
1. It's not simply a validation/syntax checker. It also checks for problematic patterns you might not be aware of.
2. It works regardless of CSS minification - it's just a parser, not a style enforcer, white space doesn't matter.

I can understand why those using a CSS preprocessor wouldn't want to use it, it's definitely hard to trace back the problem in the final CSS to the code you wrote. And if you're targeting CodeKit at only those people who use CSS preprocessors, then I totally understand not wanting to include it.

I disagree, though, that using preprocessors is the only way of "doing it right". There are a lot of good developers who don't use CSS preprocessors.

**Comment by mohsen1** [comment]
+1 for adding CSSLint.
I'm sure if CSS Lint knows my HTML it can help more efficient.

**Comment by aripalo** [comment]
+1 for CSSLint. Though it should be available as optional extra, not something that is on by default. It can be really useful tool depending on what "philosophies" one is following in his stylesheets.

**Comment by aripalo** [comment]
Well, have to comment on this how I would like to see it working:

As CSS lint can't work with LESS/SASS/etc and even if it could, the main point of it is to avoid problems / antipatterns / bad habits in outputted CSS that will be sent to browsers. The syntax checking is handled by the preprocessor compiler as it is already handled.

Of course CSS lint wouldn't match the line numbers of original LESS/SASS/etc unless you can pull out some magic tricks, but even still it would be useful. It would allow developing my stylesheet against certain rules/methodology and Codekit with CSS lint would notify me when I steer away from the CSS lint rules I've chosen. Of course one can always put the Codekit generated CSS to http://csslint.net manually every now and then, but the main reason why I want CSS lint to codekit is that it would enforce me to follow the CSS lint rules that I've chosen all the time and spot "errors" right away. It just would fit this "Codekit workflow" and the beautiful UI so well. 

How it should work then?
I'd say, when CSS lint option is checked in Codekit, then Codekit should compile LESS/SASS/etc to CSS first without minification (to some temp file or to memory or whatevs...) and then check that against CSS lint and after that has passed, then continue and create (concat/minify) the "final" outputted CSS file. Also it would be needed to allow just setting CSS lint to show warnings and still compile the final CSS file, but also allow checking a setting that'd would give "errors" if CSS lint wouldn't pass and therefore preventing the final CSS file output. Also if CSS Lint were to be added to Codekit, it should be as an optional extra.

But again, I sort of understand this isn't a mainstream feature so I can see why it is really low priority etc. So, no hard feelings if it won't get implemented :) Thanks for the Codekit, keep up the good work! Gonna buy the final version this week! :)

To recap:
- CSS Lint as an optional setting (off by default)
- default setting: show CSS Lint messages as warnings & generate the final CSS output file
- optional setting: show CSS Lint messages as errors & prevent generation of the final CSS output file
- settings panel for the CSS Lint rules (samekind as JS Hint & JS Lint have now in the settings panel)
